File: python/wginfer/models/qwen2.py
import json
from pathlib import Path
from typing import Sequence

import torch
from safetensors.torch import load_file as safetensors_load_file

from .._wginfer import DataType
from .._wginfer import DeviceType
from .._wginfer import Qwen2Meta
from .._wginfer import Qwen2Model
from ..tensor import Tensor


class Qwen2:
    def __init__(self, model_path, device: DeviceType = DeviceType.CPU):
        model_path = Path(model_path)
        self._device = device

        config_path = model_path / "config.json"
        if not config_path.exists():
            raise FileNotFoundError(f"config.json not found in {model_path}")

        with open(config_path, "r", encoding="utf-8") as f:
            config = json.load(f)

        self.meta = Qwen2Meta()
        self.meta.dtype = DataType.BF16
        self.meta.nlayer = config.get("num_hidden_layers", config.get("num_layers", 0))
        self.meta.hs = config.get("hidden_size", 0)
        self.meta.nh = config.get("num_attention_heads", 0)
        self.meta.nkvh = config.get("num_key_value_heads", self.meta.nh)
        self.meta.dh = config.get("head_dim", self.meta.hs // self.meta.nh)
        if self.meta.dh == 0:
            self.meta.dh = self.meta.hs // self.meta.nh
        self.meta.di = config.get("intermediate_size", 0)
        self.meta.maxseq = config.get("max_position_embeddings", 32768)
        self.meta.voc = config.get("vocab_size", 0)
        self.meta.epsilon = config.get("rms_norm_eps", 1e-6)
        self.meta.theta = config.get("rope_theta", 1000000.0)
        self.meta.end_token = config.get("eos_token_id", 151643)

        self.model = Qwen2Model(self.meta, device, 0)
        self._load_weights(model_path)

    def _load_weights(self, model_path):
        safetensors_files = sorted(model_path.glob("*.safetensors"))
        if not safetensors_files:
            raise FileNotFoundError(f"No safetensors files found in {model_path}")

        print(f"[wginfer] Loading Qwen2 weights from: {model_path}")
        print(f"[wginfer] Found {len(safetensors_files)} safetensors")

        def to_bf16_cpu_contig(t: torch.Tensor) -> torch.Tensor:
            t = t.detach().cpu()
            if t.dtype != torch.bfloat16:
                t = t.to(torch.bfloat16)
            return t.contiguous()

        def load_wginfer_tensor_from_torch(t: torch.Tensor) -> Tensor:
            t_cpu = to_bf16_cpu_contig(t)
            lt = Tensor(shape=list(t_cpu.shape), dtype=DataType.BF16, device=self._device)
            lt.load(t_cpu.data_ptr())
            return lt

        def set_field(name: str, t: torch.Tensor):
            self.model.set_weight(name, load_wginfer_tensor_from_torch(t))

        loaded = 0
        skipped = 0

        for file_idx, file in enumerate(safetensors_files):
            print(f"[wginfer] [{file_idx + 1}/{len(safetensors_files)}] reading {file.name}")
            weights_dict = safetensors_load_file(str(file))
            print(f"[wginfer]   tensors in shard: {len(weights_dict)}")

            for key, t in weights_dict.items():
                if key == "model.embed_tokens.weight":
                    set_field("in_embed", t)
                    loaded += 1
                    continue
                if key == "lm_head.weight":
                    set_field("out_embed", t)
                    loaded += 1
                    continue
                if key == "model.norm.weight":
                    set_field("out_norm_w", t)
                    loaded += 1
                    continue

                if not key.startswith("model.layers."):
                    skipped += 1
                    continue

                parts = key.split(".")
                if len(parts) < 4:
                    skipped += 1
                    continue

                try:
                    layer_idx = int(parts[2])
                except ValueError:
                    skipped += 1
                    continue

                if layer_idx < 0 or layer_idx >= int(self.meta.nlayer):
                    skipped += 1
                    continue

                suffix = ".".join(parts[3:])
                lt = load_wginfer_tensor_from_torch(t)

                if suffix == "input_layernorm.weight":
                    self.model.set_layer_weight("attn_norm_w", layer_idx, lt)
                    loaded += 1
                    continue
                if suffix == "self_attn.q_proj.weight":
                    self.model.set_layer_weight("attn_q_w", layer_idx, lt)
                    loaded += 1
                    continue
                if suffix == "self_attn.q_proj.bias":
                    self.model.set_layer_weight("attn_q_b", layer_idx, lt)
                    loaded += 1
                    continue
                if suffix == "self_attn.k_proj.weight":
                    self.model.set_layer_weight("attn_k_w", layer_idx, lt)
                    loaded += 1
                    continue
                if suffix == "self_attn.k_proj.bias":
                    self.model.set_layer_weight("attn_k_b", layer_idx, lt)
                    loaded += 1
                    continue
                if suffix == "self_attn.v_proj.weight":
                    self.model.set_layer_weight("attn_v_w", layer_idx, lt)
                    loaded += 1
                    continue
                if suffix == "self_attn.v_proj.bias":
                    self.model.set_layer_weight("attn_v_b", layer_idx, lt)
                    loaded += 1
                    continue
                if suffix == "self_attn.o_proj.weight":
                    self.model.set_layer_weight("attn_o_w", layer_idx, lt)
                    loaded += 1
                    continue
                if suffix == "post_attention_layernorm.weight":
                    self.model.set_layer_weight("mlp_norm_w", layer_idx, lt)
                    loaded += 1
                    continue
                if suffix == "mlp.gate_proj.weight":
                    self.model.set_layer_weight("mlp_gate_w", layer_idx, lt)
                    loaded += 1
                    continue
                if suffix == "mlp.up_proj.weight":
                    self.model.set_layer_weight("mlp_up_w", layer_idx, lt)
                    loaded += 1
                    continue
                if suffix == "mlp.down_proj.weight":
                    self.model.set_layer_weight("mlp_down_w", layer_idx, lt)
                    loaded += 1
                    continue

                skipped += 1

            del weights_dict

        print(f"[wginfer] Done. loaded={loaded}, skipped={skipped}")

    def generate(
        self,
        inputs: Sequence[int],
        max_new_tokens: int = None,
        top_k: int = 1,
        top_p: float = 0.8,
        temperature: float = 0.8,
    ):
        self.model.reset_cache()

        output_tokens = list(inputs)
        if len(inputs) == 0:
            return output_tokens

        if max_new_tokens is None:
            max_new_tokens = 128
        max_new_tokens = max(int(max_new_tokens), 1)

        next_token = self._infer_next(inputs, top_k, top_p, temperature)
        output_tokens.append(next_token)

        for _ in range(max_new_tokens - 1):
            if next_token == self.meta.end_token:
                break
            next_token = self._infer_next([next_token], top_k, top_p, temperature)
            output_tokens.append(next_token)

        return output_tokens

    def generate_stream(
        self,
        inputs: Sequence[int],
        max_new_tokens: int = None,
        top_k: int = 1,
        top_p: float = 1.0,
        temperature: float = 1.0,
    ):
        self.model.reset_cache()
        if len(inputs) == 0:
            return

        if max_new_tokens is None:
            max_new_tokens = 128
        max_new_tokens = max(int(max_new_tokens), 1)

        next_token = self._infer_next(inputs, top_k, top_p, temperature)
        yield next_token
        for _ in range(max_new_tokens - 1):
            if next_token == self.meta.end_token:
                break
            next_token = self._infer_next([next_token], top_k, top_p, temperature)
            yield next_token

    def _infer_next(
        self,
        tokens: Sequence[int],
        top_k: int,
        top_p: float,
        temperature: float,
    ) -> int:
        return int(
            self.model.infer(
                list(tokens),
                int(top_k),
                float(top_p),
                float(temperature),
            )
        )

File: python/bindings/models/qwen2.cpp
#include "bindings.hpp"
#include "../core/core.hpp"

#include <pybind11/stl.h>

#include "models/qwen2/model.hpp"

#include <memory>
#include <stdexcept>
#include <string>
#include <vector>

namespace py = pybind11;
namespace qwen2 = wginfer::models::qwen2;

namespace wginfer::pybind {

namespace {

void assign_global_weight(
    qwen2::ModelWeights &weights,
    const std::string &name,
    const wginfer::tensor_t &tensor) {
    if (name == "in_embed") {
        weights.in_embed = tensor;
        return;
    }
    if (name == "out_embed") {
        weights.out_embed = tensor;
        return;
    }
    if (name == "out_norm_w") {
        weights.out_norm_w = tensor;
        return;
    }
    throw std::invalid_argument("Unknown global Qwen2 weight: " + name);
}

void assign_layer_weight(
    qwen2::ModelWeights &weights,
    const std::string &name,
    size_t layer_idx,
    const wginfer::tensor_t &tensor) {
    if (name == "attn_norm_w") {
        weights.attn_norm_w.at(layer_idx) = tensor;
        return;
    }
    if (name == "attn_q_w") {
        weights.attn_q_w.at(layer_idx) = tensor;
        return;
    }
    if (name == "attn_q_b") {
        weights.attn_q_b.at(layer_idx) = tensor;
        return;
    }
    if (name == "attn_k_w") {
        weights.attn_k_w.at(layer_idx) = tensor;
        return;
    }
    if (name == "attn_k_b") {
        weights.attn_k_b.at(layer_idx) = tensor;
        return;
    }
    if (name == "attn_v_w") {
        weights.attn_v_w.at(layer_idx) = tensor;
        return;
    }
    if (name == "attn_v_b") {
        weights.attn_v_b.at(layer_idx) = tensor;
        return;
    }
    if (name == "attn_o_w") {
        weights.attn_o_w.at(layer_idx) = tensor;
        return;
    }
    if (name == "mlp_norm_w") {
        weights.mlp_norm_w.at(layer_idx) = tensor;
        return;
    }
    if (name == "mlp_gate_w") {
        weights.mlp_gate_w.at(layer_idx) = tensor;
        return;
    }
    if (name == "mlp_up_w") {
        weights.mlp_up_w.at(layer_idx) = tensor;
        return;
    }
    if (name == "mlp_down_w") {
        weights.mlp_down_w.at(layer_idx) = tensor;
        return;
    }
    throw std::invalid_argument("Unknown per-layer Qwen2 weight: " + name);
}

class PyQwen2Model {
public:
    PyQwen2Model(const qwen2::ModelMeta &meta, wginferDeviceType_t device, int device_id)
        : model_(std::make_unique<qwen2::Model>(meta, device, device_id)) {
    }

    const qwen2::ModelMeta &meta() const {
        return model_->meta();
    }

    void reset_cache() const {
        model_->reset_cache();
    }

    void set_weight(const std::string &name, const std::shared_ptr<PyTensor> &tensor) {
        assign_global_weight(model_->weights(), name, tensor->tensor());
    }

    void set_layer_weight(const std::string &name, size_t layer_idx, const std::shared_ptr<PyTensor> &tensor) {
        if (layer_idx >= model_->meta().nlayer) {
            throw std::out_of_range("Qwen2 layer_idx out of range");
        }
        assign_layer_weight(model_->weights(), name, layer_idx, tensor->tensor());
    }

    int64_t infer(
        const std::vector<int64_t> &token_ids,
        int top_k,
        float top_p,
        float temperature) const {
        std::vector<int64_t> tokens = token_ids;
        return model_->infer(
            tokens.data(),
            tokens.size(),
            top_k,
            top_p,
            temperature);
    }

private:
    std::unique_ptr<qwen2::Model> model_;
};

} // namespace

void bind_qwen2(py::module_ &m) {
    py::class_<qwen2::ModelMeta>(m, "Qwen2Meta")
        .def(py::init<>())
        .def_readwrite("dtype", &qwen2::ModelMeta::dtype)
        .def_readwrite("nlayer", &qwen2::ModelMeta::nlayer)
        .def_readwrite("hs", &qwen2::ModelMeta::hs)
        .def_readwrite("nh", &qwen2::ModelMeta::nh)
        .def_readwrite("nkvh", &qwen2::ModelMeta::nkvh)
        .def_readwrite("dh", &qwen2::ModelMeta::dh)
        .def_readwrite("di", &qwen2::ModelMeta::di)
        .def_readwrite("maxseq", &qwen2::ModelMeta::maxseq)
        .def_readwrite("voc", &qwen2::ModelMeta::voc)
        .def_readwrite("epsilon", &qwen2::ModelMeta::epsilon)
        .def_readwrite("theta", &qwen2::ModelMeta::theta)
        .def_readwrite("end_token", &qwen2::ModelMeta::end_token);

    py::class_<PyQwen2Model>(m, "Qwen2Model")
        .def(
            py::init<const qwen2::ModelMeta &, wginferDeviceType_t, int>(),
            py::arg("meta"),
            py::arg("device"),
            py::arg("device_id") = 0)
        .def("meta", &PyQwen2Model::meta, py::return_value_policy::reference_internal)
        .def("reset_cache", &PyQwen2Model::reset_cache)
        .def("set_weight", &PyQwen2Model::set_weight, py::arg("name"), py::arg("tensor"))
        .def("set_layer_weight", &PyQwen2Model::set_layer_weight, py::arg("name"), py::arg("layer_idx"), py::arg("tensor"))
        .def(
            "infer",
            &PyQwen2Model::infer,
            py::arg("token_ids"),
            py::arg("top_k") = 1,
            py::arg("top_p") = 1.0f,
            py::arg("temperature") = 1.0f);
}

} // namespace wginfer::pybind
